src/Simplex/Messaging/Agent.hs

```diff
@@ -865,19 +873,38 @@ getNotificationMessage' c nonce encNtfInfo = do
           Just SMP.NMsgMeta {msgId = msgId', msgTs = msgTs'} -> msgId == msgId' || msgTs > msgTs'
           Nothing -> SMP.notification msgFlags
 
+type EIORef a = IORef (Either AgentErrorType a)
+
 -- | Send message to the connection (SEND command) in Reader monad
 sendMessage' :: forall m. AgentMonad m => AgentClient -> ConnId -> MsgFlags -> MsgBody -> m AgentMsgId
-sendMessage' c connId msgFlags msg = withConnLock c connId "sendMessage" $ do
-  SomeConn _ conn <- withStore c (`getConn` connId)
-  case conn of
-    DuplexConnection cData _ sqs -> enqueueMsgs cData sqs
-    SndConnection cData sq -> enqueueMsgs cData [sq]
-    _ -> throwError $ CONN SIMPLEX
+sendMessage' c connId msgFlags msg =
+  oneResult $ \r -> sendMessagesB c [(r, (connId, msgFlags, msg))]
+
+-- | Send multiple messages to different connections (SEND command) in Reader monad
+sendMessages' :: forall m. AgentMonad' m => AgentClient -> [MsgReq] -> m [Either AgentErrorType AgentMsgId]
+sendMessages' c msgReqs = do
+  rs <- replicateM (length msgReqs) (newIORef $ Left $ INTERNAL "skipped in batch")
+  sendMessagesB c $ zip rs msgReqs
+  mapM readIORef rs
+
+sendMessagesB :: forall m. AgentMonad' m => AgentClient -> [(EIORef AgentMsgId, MsgReq)] -> m ()
+sendMessagesB c reqs = withConnLocks c connIds "sendMessages" $ do
+  reqs' <- zip reqs <$> withStoreBatch c (\db -> map (getConn db) connIds)
+  reqs'' <- catMaybes <$> mapM prepareConn reqs'
+  enqueueMessagesB c reqs''
   where
-    enqueueMsgs :: ConnData -> NonEmpty SndQueue -> m AgentMsgId
-    enqueueMsgs cData sqs = do
-      when (ratchetSyncSendProhibited cData) $ throwError $ CMD PROHIBITED
-      enqueueMessages c cData sqs msgFlags $ A_MSG msg
+    prepareConn :: ((EIORef AgentMsgId, MsgReq), Either AgentErrorType SomeConn) -> m (Maybe (EIORef AgentMsgId, (ConnData, NonEmpty SndQueue, MsgFlags, AMessage)))
+    prepareConn (req@(r, _), conn_) = case conn_ of
+      Left e -> Nothing <$ writeIORef r (Left e)
+      Right (SomeConn _ conn) -> case conn of
+        DuplexConnection cData _ sqs -> enqueueMsgs cData sqs req
+        SndConnection cData sq -> enqueueMsgs cData [sq] req
+        _ -> Nothing <$ writeIORef r (Left $ CONN SIMPLEX)
+    enqueueMsgs :: ConnData -> NonEmpty SndQueue -> (EIORef AgentMsgId, MsgReq) -> m (Maybe (EIORef AgentMsgId, (ConnData, NonEmpty SndQueue, MsgFlags, AMessage)))
```

Comment:
name is confusing, it's almost the same as function that actually does enqueue messages, but this one doesn't.

rename to something like `msgsToEnqueue` / `prepareEnqueuedMsgs`

Reply:
or just `prepareMsg` similar to `prepareConn`, as it actually prepares a single message, not messages